### PR TITLE
Add Registry DSL

### LINKED ISSUE #5: Registration API/DSL
## Intro:
Registering items, biomes, blocks, ect. in Minecraft mostly makes use of a line similar to `Registry.register(Registry.TYPE, new Identifier("modid", "some_path"), actual_obj)`, this can become very repetitive and hard to read through when repeated a significant amount. A simple API or DSL system could reduce a lot of this repetition or other crust.

## Existing work:
- At this time I am unable to locate existing work on such an API

### PR DESCRIPTION
**Pull request by Oliver-makes-code**
This PR provides a Registry DSL for organizing registration
Example:
```kt
public val TEST_BLOCK: Block = TODO()
private fun test() {
    RegistryDsl("modid") {
        Registry.BLOCK {
            TEST_BLOCK withName "test_3"
        }
        TEST_BLOCK withName "test_2" toRegistry Registry.BLOCK
    }
    Registry.BLOCK("modid") {
        TEST_BLOCK withName "test_3"
    }
    Registry.BLOCK {
        TEST_BLOCK withId Identifier("modid","test_4")
    }
    TEST_BLOCK withId Identifier("modid", "test_5") toRegistry Registry.BLOCK
}
```

Closes #5

### CODE REVIEW

**Comment by Oliver-makes-code** [comment]
It's also important to note that it works with all registries, not just BLOCK

**Comment by Oliver-makes-code** [comment]
> I know this would increase api complexity but... Would it be possible to have the following
> 
> ```kotlin
> Registry.BLOCK {
>     Block(/*...*/) withName "block_1"
> }
> ```
> 
> instead of
> 
> ```kotlin
> Registry.BLOCK {
>     Block(/*...*/) withName "block_1" toRegistry this
> }
> ```

Currently, that can only been done with `context`, which is experimental right now

**Comment by Kroppeb** [comment]
> Currently, that can only been done with `context`, which is experimental right now

Mmh, I feel like this is possible without `context`

**Comment by sschr15** [comment]
> Mmh, I feel like this is possible without `context`

In that particular case, I believe (though I haven't looked deeply into it yet) that it is possible, but in some other scenarios, context receivers are very handy

**Comment by Oliver-makes-code** [comment]
> > Mmh, I feel like this is possible without `context`
> 
> In that particular case, I believe (though I haven't looked deeply into it yet) that it is possible, but in some other scenarios, context receivers are very handy

How would I go about doing that then?

**Comment by Kroppeb** [comment]
> How would I go about doing that then?

time to clone, and write some actual kotlin code for the first time in too long

**Comment by sschr15** [comment]
`Registry.() -> Unit` or a similar signature for the passed parameter

**Comment by sschr15** [comment]
Oh I already see that's what it is; in that case it should already work

**Comment by sschr15** [comment]
oh wait I completely misunderstood it all

**Comment by SilverAndro** [comment]
[I dont think it is possible?](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS42LjIxIiwiY29kZSI6ImNsYXNzIFNvbWVUeXBlIHtcbiAgICBpbmZpeCBmdW4gZG9UaGluZyhvdGhlcjogSW50KSB7XG4gICAgICAgIHByaW50bG4oXCJBICRvdGhlciAkdGhpc1wiKVxuICAgIH1cbn1cblxuY2xhc3MgU29tZVdyYXBwZXIoYWN0aW9uOiBTb21lV3JhcHBlci4oKS0+VW5pdCkge1xuICAgIGluaXQge1xuICAgICAgICBhY3Rpb24oKVxuICAgIH1cbiAgICBcbiAgICBpbmZpeCBmdW4gU29tZVR5cGUuZG9UaGluZyhvdGhlcjogSW50KSB7XG4gICAgICAgIHByaW50bG4oXCJCICRvdGhlciAkdGhpc1wiKVxuICAgIH1cbiAgICBcbiAgICBpbmZpeCBmdW4gU29tZVR5cGUuZG9PdGhlclRoaW5nKG90aGVyOiBJbnQpIHtcbiAgICAgICAgcHJpbnRsbihcIkMgJG90aGVyICR0aGlzXCIpXG4gICAgfVxufVxuXG5mdW4gbWFpbigpIHtcbiAgICBTb21lVHlwZSgpIGRvVGhpbmcgMVxuICAgIFNvbWVXcmFwcGVyIHtcbiAgICAgICAgU29tZVR5cGUoKSBkb1RoaW5nIDJcbiAgICAgICAgU29tZVR5cGUoKSBkb090aGVyVGhpbmcgM1xuICAgICAgICAvLyBJbnZhbGlkOiBTb21lVHlwZSgpIHRoaXMuZG9UaGluZyAyXG4gICAgfVxufSIsInBsYXRmb3JtIjoiamF2YSIsImFyZ3MiOiIifQ==) Cant find a way to access the wrapper one if its the same, and even then the infix function ends up overwriting `this`, meaning you'd have to manage to hack around that by making everything run on one `object`, which wouldn't be thread safe.

**Comment by sschr15** [comment]
The only reasonable option that I could see to remove the terminating call would be to make every call terminating, which would not work well if attributes get thrown in as well

**Comment by sschr15** [comment]
That or a temporary state could be kept, and as soon as the state is no longer touched, it's automatically added to the registry

**Comment by Oliver-makes-code** [comment]
Edited the example in the first comment to match changes made in the latest commit

**Comment by Oliver-makes-code** [comment]
I think how it is currently is its final form, so once all 3 requested reviewers review positively I think it should be safe to merge

File: wrapper/main/src/main/kotlin/org/quiltmc/qkl/wrapper/main/registry/RegistryDsl.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     *
+     * @author Oliver-makes-code (Emma)
+     * */
+    public infix fun toRegistry(registry: Registry<T>): T {
+        return Registry.register(registry, Identifier(modid, name), this.t)
+    }
+}
+
+/**
+ * Represent an action with a registry
```
Comment (SilverAndro) [inline wrapper/main/src/main/kotlin/org/quiltmc/qkl/wrapper/main/registry/RegistryDsl.kt:27]: Small nitpick: Should probably be `Represents`

File: wrapper/main/src/main/kotlin/org/quiltmc/qkl/wrapper/main/registry/RegistryDsl.kt [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        if (modid == null) {
+            Registry.register(registry, Identifier(name), this)
+            return
+        }
+        Registry.register(registry, Identifier(modid, name), this)
+    }
+}
+
+/**
+ * Registry DSL class, used to register object more cleanly
```
Comment (SilverAndro) [inline wrapper/main/src/main/kotlin/org/quiltmc/qkl/wrapper/main/registry/RegistryDsl.kt:60]: Small nitpick: Should probably be `objects`

File: wrapper/main/src/main/kotlin/org/quiltmc/qkl/wrapper/main/registry/RegistryDsl.kt [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+     * @param id The Identifier to register under
+     *
+     * @author Oliver-makes-code (Emma)
+     * */
+    public infix fun T.withId(id: Identifier) {
+        Registry.register(registry, id, this)
+    }
+
+    /**
+     * Registers an object with a given name
```
Comment (SilverAndro) [inline wrapper/main/src/main/kotlin/org/quiltmc/qkl/wrapper/main/registry/RegistryDsl.kt:45]: Probably worth adding its behavior with/without a modid, i.e make it clear how the `name`  param is used

**Review by SilverAndro — CHANGES_REQUESTED** [review]
Looks good just a few documentation nitpicks

**Review by SilverAndro — APPROVED** [review]

**Review by PotatoPresident — APPROVED** [review]
Looks good to me

**Review by leocth — APPROVED** [review]
LGTM, but could we consider the possibility of having a DeferredRegistry-style DSL, as mentioned by Silver in Discord?

```kt
object MyContent : Register("modid") {
  val SOME_ITEM by item("magic", ItemInstance())
  val SOME_BLOCK by block("magic_block", BlockInstance())
}
```

**Comment by Oliver-makes-code** [comment]
> ```kotlin
> object MyContent : Register("modid") {
>   val SOME_ITEM by item("magic", ItemInstance())
>   val SOME_BLOCK by block("magic_block", BlockInstance())
> }
> ```

The issue with this kind of system is that it doesn't have automatic support with third party registries

**Comment by leocth** [comment]
> > ```kotlin
> > object MyContent : Register("modid") {
> >   val SOME_ITEM by item("magic", ItemInstance())
> >   val SOME_BLOCK by block("magic_block", BlockInstance())
> > }
> > ```
> 
> The issue with this kind of system is that it doesn't have automatic support with third party registries

I mean, you could just add extension methods to the `Register` class but I do see your point - i just think it's a bit neater to have the item declaration, initialization and registration in one place

**Review by NoComment1105 — APPROVED** [review]
LGTM 👍🏼

**Review by Leo40Git — APPROVED** [review]

**Pull request merged**